btcec/schnorr.go

```diff
@@ -0,0 +1,333 @@
+
+	// Negate k if y(R) is odd.
+	if !hasEvenY(Ry) {
+		k.Sub(n, k)
+	}
+
+	// e = int(hashBIP0340/challenge(R || P || m)) mod n
+	{
+		m := make([]byte, 96)
+		copy(m[:32], intToBytes(32, Rx))
+		copy(m[32:64], intToBytes(32, Px))
+		copy(m[64:96], msg)
+		e.SetBytes(taggedHash(BIP340Challenge, m))
+		e.Mod(e, n)
+	}
+
+	// (k + ed) mod n
+	s.Mul(e, d)
+	s.Add(k, s)
+	s.Mod(s, n)
+
+	// Signature is (x(R), s).
+	copy(sig[:32], intToBytes(32, Rx))
+	copy(sig[32:], intToBytes(32, s))
+
+	// Verify signature before returning.
+	if verify, err := schnorrVerify(msg, intToBytes(32, Px), sig[:]); !verify || err != nil {
+		return sig, fmt.Errorf("cannot create signature: %w", err)
+	}
+
+	return sig, nil
+}
+
+func liftX(key []byte) (*big.Int, *big.Int, error) {
+	x := new(big.Int).SetBytes(key)
+
+	// p is field size.
+	p := S256().P
+
+	if x.Cmp(p) >= 0 {
+		return nil, nil, errors.New("inf")
+	}
+
+	// c = x^3 + 7 mod P.
+	c := new(big.Int)
+	c.Exp(x, three, p)
+	c.Add(c, seven)
+	c.Mod(c, p)
+
+	// y = c^((p+1)/4) mod P.
+	y := new(big.Int)
+	y.Add(p, one)
+	y.Div(y, four)
+	y.Exp(c, y, p)
+
+	if new(big.Int).And(y, one).Cmp(zero) != 0 {
+		y.Sub(p, y)
+	}
+
+	return x, y, nil
+}
+
+func schnorrVerify(msg, publicKey, signature []byte) (bool, error) {
+	if l := len(msg); l != schnorrMessageLen {
+		return false, fmt.Errorf("message is not 32 bytes : got %d, want %d", l, schnorrMessageLen)
+	}
+
+	if l := len(publicKey); l != schnorrPublicKeyLen {
+		return false, fmt.Errorf("public key is not 32 bytes : got %d, want %d", l, schnorrPublicKeyLen)
+	}
+
+	if l := len(signature); l != schnorrSignatureLen {
+		return false, fmt.Errorf("signature is not 32 bytes : got %d, want %d", l, schnorrSignatureLen)
+	}
+
+	curve := S256()
+
+	// n is the curve order.
+	n := curve.N
+
+	p := curve.P
+
+	r := new(big.Int)
+	s := new(big.Int)
+
+	e := new(big.Int)
+
+	// Get Point P from the x-coordinate.
+	Px, Py, err := liftX(publicKey[:])
+	if err != nil {
+		return false, err
+	}
+
+	// Check that P is on the curve.
+	if !curve.IsOnCurve(Px, Py) {
+		return false, errors.New("point P is not on the curve")
+	}
+
+	r.SetBytes(signature[:32])
+	s.SetBytes(signature[32:])
+
+	// Fail if s >= n
+	if s.Cmp(n) >= 0 {
+		return false, errors.New("s is greater than curve order")
+	}
+
+	// Fail if r >= p
+	if r.Cmp(p) >= 0 {
+		return false, errors.New("r is greater than p")
+	}
+
+	// e = sha256(hashBIP0340/challenge || r || P || m) mod n.
+	{
+		m := make([]byte, 96)
+		copy(m[:32], signature[:32])
+		copy(m[32:64], publicKey[:])
+		copy(m[64:], msg)
+		e.SetBytes(taggedHash(BIP340Challenge, m))
+		e.Mod(e, n)
+	}
+
+	// s*G
+	sGx, sGy := curve.ScalarBaseMult(s.Bytes())
+
+	// (n - e)*P
+	ePx, ePy := curve.ScalarMult(Px, Py, new(big.Int).Sub(n, e).Bytes())
+
+	// R = s*G + (N-e)*P
+	Rx, Ry := curve.Add(sGx, sGy, ePx, ePy)
+
+	// Fail if R is at infinity.
+	if Rx.Cmp(zero) == 0 || Ry.Cmp(zero) == 0 {
+		return false, errors.New("point R is at infinity")
```

Comment:
instead of constructing new errors on every failure (you would recognize them only by the string), why not give them top-level names and make them part of the interface? then the tests can assert that the correct error is returned.